expression/constant_propagation.go

```diff
@@ -290,37 +306,276 @@ func (s *propagateConstantSolver) solve(conditions []Expression) []Expression {
 	}
 	s.propagateConstantEQ()
 	s.propagateColumnEQ()
-	for i, cond := range s.conditions {
-		if dnf, ok := cond.(*ScalarFunction); ok && dnf.FuncName.L == ast.LogicOr {
-			dnfItems := SplitDNFItems(cond)
-			for j, item := range dnfItems {
-				dnfItems[j] = ComposeCNFCondition(s.ctx, PropagateConstant(s.ctx, []Expression{item})...)
+	s.conditions = propagateConstantDNF(s.ctx, s.conditions)
+	return s.conditions
+}
+
+// PropagateConstant propagate constant values of deterministic predicates in a condition.
+func PropagateConstant(ctx sessionctx.Context, conditions []Expression) []Expression {
+	solver := &propConstSolver{}
+	solver.colMapper = make(map[int64]int)
+	solver.ctx = ctx
+	return solver.solve(conditions)
+}
+
+type propOuterJoinConstSolver struct {
+	basePropConstSolver
+	joinConds   []Expression
+	filterConds []Expression
+	outerSchema *Schema
+	innerSchema *Schema
+}
+
+func (s *propOuterJoinConstSolver) setConds2ConstFalse(joinConds, filterConds bool) {
```

Comment:
maybe the first parameter can be removed, because it's always `true`:
```
  5 expression/constant_propagation.go|367 col 9| s.setConds2ConstFalse(true, true)
  6 expression/constant_propagation.go|369 col 9| s.setConds2ConstFalse(true, false)
  7 expression/constant_propagation.go|384 col 7| s.setConds2ConstFalse(true, true)
  8 expression/constant_propagation.go|386 col 7| s.setConds2ConstFalse(true, false)
```